### PR TITLE
Firebase Hosting Custom Domains

### PR DESCRIPTION
**Pull request by rainshen49**
Add support for adding a custom domain with Firebase Hosting.
https://github.com/hashicorp/terraform-provider-google/issues/12955

If this PR is for Terraform, I acknowledge that I have:

- [x] Searched through the [issue tracker](https://github.com/hashicorp/terraform-provider-google/issues) for an open issue that this either resolves or contributes to, commented on it to claim it, and written "fixes {url}" or "part of {url}" in this PR description. If there were no relevant open issues, I opened one and commented that I would like to work on it (not necessary for very small changes).
- [x] Ensured that all new fields I added that can be set by a user appear in at least one [example](https://github.com/GoogleCloudPlatform/magic-modules/tree/main/mmv1/templates/terraform/examples) (for generated resources) or [third_party test](https://github.com/GoogleCloudPlatform/magic-modules/tree/main/mmv1/third_party/terraform/tests) (for handwritten resources or update tests).
- [x] [Generated Terraform providers](https://github.com/GoogleCloudPlatform/magic-modules/blob/main/docs/content/docs/getting-started/generate-providers.md), and ran [`make test` and `make lint`](https://googlecloudplatform.github.io/magic-modules/docs/getting-started/run-provider-tests/#run-unit-tests) in the generated providers to ensure it passes unit and linter tests.
- [x] [Ran](https://github.com/GoogleCloudPlatform/magic-modules/blob/main/docs/content/docs/getting-started/run-provider-tests.md) relevant acceptance tests using my own Google Cloud project and credentials (If the acceptance tests do not yet pass or you are unable to run them, please let your reviewer know).
- [x] Read [Write release notes](https://googlecloudplatform.github.io/magic-modules/contribute/release-notes/) before writing my release note below.

**Release Note Template for Downstream PRs (will be copied)**

```release-note:new-resource
`google_firebase_hosting_custom_domain`
```

### CODE REVIEW

**Comment by rainshen49** [comment]
NOTE to reviewer: looking for some early code review feedback. The tests will fail because the underlying API is pending final approval before opening to public.

File: mmv1/products/firebasehosting/CustomDomain.yaml
Comment (roaks3) [inline mmv1/products/firebasehosting/CustomDomain.yaml:30]: Note to make sure this URL is reachable once the API is public

File: mmv1/products/firebasehosting/CustomDomain.yaml
Comment (roaks3) [inline mmv1/products/firebasehosting/CustomDomain.yaml:62]: Nit: I believe this is implied by the `min_version` at the resource level. Unless you expect some of these examples to stay beta when the resource is promoted to GA, you should be able to remove this line.

File: mmv1/products/firebasehosting/CustomDomain.yaml
Comment (roaks3) [inline mmv1/products/firebasehosting/CustomDomain.yaml:119]: Could you clarify why this is a parameter instead of a property?

File: mmv1/products/firebasehosting/CustomDomain.yaml
Comment (roaks3) [inline mmv1/products/firebasehosting/CustomDomain.yaml:275]: I don't believe the `output: true` behavior is inherited, so you may need to add it to each of the children here. Were you able to confirm that values like `check_time` cannot be specified in config?

File: mmv1/templates/terraform/post_create/firebasehosting_custom_domain_wait_dns_verification.erb
Comment (roaks3) [inline mmv1/templates/terraform/post_create/firebasehosting_custom_domain_wait_dns_verification.erb:1]: This custom code might be worth a brief comment

**Review by roaks3 — COMMENTED** [review]
Just did an initial review. Overall things look good, although I didn't look too closely at the tests yet. Only a few small comments.

File: mmv1/products/firebasehosting/CustomDomain.yaml
Comment (rainshen49) [reply mmv1/products/firebasehosting/CustomDomain.yaml:275]: Yeah. Here's what I did
```
  required_dns_updates {
    check_time = "12345"
  }
```
And here's the error I got
`Can't configure a value for "required_dns_updates": its value will be decided automatically based on the result of applying this configuration.`

File: mmv1/products/firebasehosting/CustomDomain.yaml
Comment (rainshen49) [reply mmv1/products/firebasehosting/CustomDomain.yaml:119]: This field is not part of the CustomDomain entity. It is used to control the creation behavior of Terraform whether to wait for DNS verification before considering this resource as created. IIUC, this is a use case for a parameter, right?

File: mmv1/products/firebasehosting/CustomDomain.yaml
Comment (rainshen49) [reply mmv1/products/firebasehosting/CustomDomain.yaml:62]: Done.

File: mmv1/products/firebasehosting/CustomDomain.yaml
Comment (roaks3) [reply mmv1/products/firebasehosting/CustomDomain.yaml:116]: Had to look it up myself, and came across [`virtual_fields`](https://github.com/GoogleCloudPlatform/magic-modules/blob/main/mmv1/provider/terraform/virtual_fields.rb). I think you may want to try that instead, as it mentions you can get similar behavior with `url_param_only`, but semantically it seems like `virtual_fields` is a better match for what you are doing.

If for some reason that won't work, it seems like what you have here is probably sufficient.

File: mmv1/products/firebasehosting/CustomDomain.yaml
Comment (roaks3) [reply mmv1/products/firebasehosting/CustomDomain.yaml:272]: 👍 thanks for checking!

File: mmv1/products/firebasehosting/CustomDomain.yaml
Comment (rainshen49) [reply mmv1/products/firebasehosting/CustomDomain.yaml:119]: Done

File: mmv1/products/firebasehosting/CustomDomain.yaml
Comment (rainshen49) [reply mmv1/products/firebasehosting/CustomDomain.yaml:62]: Update: I have to add this back or else I got this error:

> github.com/GoogleCloudPlatform/magic-modules/mmv1/api/product.rb:139:in `version_obj_or_closest': Could not find object for version ga and product Firebase Hosting (RuntimeError)`

**Comment by gustavovalverde** [comment]
I'd love to have this available to do our whole deployment available in terraform :)

File: mmv1/products/firebasehosting/CustomDomain.yaml
Comment (rainshen49) [reply mmv1/products/firebasehosting/CustomDomain.yaml:30]: Done.

**Comment by roaks3** [comment]
It looks like this is getting close to completion, and I will be on PTO in the near-term, so re-rolling the reviewer.

Note to the next reviewer: I did give this a review previously while it was in draft, and all LGTM. There might have been other changes since then, so it is probably worth another look, but hopefully nothing major.

**Comment by rainshen49** [comment]
Yeah I've been working on getting the tests to pass. Will get it off draft when ready.

**Comment by rainshen49** [comment]
@SarahFrench This PR is ready for review! I'm going to be OOO tomorrow until Sep 24. For quick questions, pls check with @tylerg-dev or @mraouffouad

**Comment by rainshen49** [comment]
@roaks3 Friendly ping

File: mmv1/products/firebasehosting/CustomDomain.yaml
Comment (roaks3) [inline mmv1/products/firebasehosting/CustomDomain.yaml:160]: Nit: I didn't mention this before, but you may want to consider changing these output-only enums to simple strings. The validation shouldn't be necessary, and a string would be more flexible to server-side changes.

**Review by roaks3 — APPROVED** [review]
LGTM, just had one nit.

Also, I assume you already considered this, but is there any way we can test with `wait_dns_verification = true`?

File: mmv1/products/firebasehosting/CustomDomain.yaml
Comment (rainshen49) [reply mmv1/products/firebasehosting/CustomDomain.yaml:160]: Done.

**Comment by rainshen49** [comment]
Yes, if using `wait_dns_verification = true`
1. It would wait for the domain to be ready, which usually takes 24 hours and time out for the tests, or
2. It would need a pre-validated real domain, and only one hosting site can be associated with a custom domain. This can become cumbersome to manage since multiple instances of the acceptance tests can run simultaneously.

**Review by roaks3 — APPROVED** [review]

**Comment by roaks3** [comment]
Merging is locked at the moment, but this can be merged once the lock is released

**Pull request merged**
